File: 2015/utils/io.py
import io


def print_solution(solution):
    print("\n=============================================================")
    print("The solution is:")
    print(solution)


def read_as_string(challenge):
    with open_challenge_file(challenge) as f:
        return f.read()


def read_by_character(challenge):
    with open_challenge_file(challenge) as f:
        while True:
            c = f.read(1)
            if not c:
                break
            yield c


def read_by_line(challenge):
    with open_challenge_file(challenge) as f:
        while True:
            line = f.readline()
            if not line:
                break
            yield line


def create_path_for_challenge(challenge):
    return f".\\inputs\\{challenge}.txt"


def open_challenge_file(challenge):
    path = create_path_for_challenge(challenge)
    return open(path, 'r')

File: fsharp/AdventOfCode.2021/AdventOfCode.2021/Utils/IO.fs
module Utils.IO

open System
open System.IO

let InputFolder = "Inputs"

let private challenge2Filename challenge =
    Path.Combine [| InputFolder; sprintf "%02d.txt" challenge |]

let readLines<'T> (challenge:int) =
    let filename = challenge2Filename challenge
    File.ReadAllLines(filename)
        |> Seq.map (fun x -> Convert.ChangeType(x, typeof<'T>) :?> 'T)

let readLine<'T> (challenge:int) (separator:char) =
    let filename = challenge2Filename challenge
    File.ReadAllText (filename)
    |> (fun line -> line.Split(separator))
    |> Seq.map (fun s -> Convert.ChangeType(s, typeof<'T>) :?> 'T)